Added support for logsafe Arg arrays in Slf4jLogsafeArgs.

aioobe opened the pull request:
## Before this PR
`Slf4jLogsafeArgs` complains that an `Arg<?>[]` isn't a logsafe argument.

## After this PR
==COMMIT_MSG==
Slf4jLogsafeArgs now allows Arg<?>[] to be passed as vararg argument to logging methods.
==COMMIT_MSG==

## Possible downsides?
None.

## Context
Only allowing `Arg<?>[]` gets us half way. Compiler would still give the the following diagnostic warning:

```
... non-varargs call of varargs method with inexact argument type for last parameter;
        cast to java.lang.Object for a varargs call
        cast to java.lang.Object[] for a non-varargs call and to suppress this warning 
```

By following the above suggestion and casting to `Object[]` the expression is no longer of `Arg<?>[]` type, hence we add support also for `Arg<?>[]` cast into an `Object[]`.

For `// BUG: Diagnostic contains: varargs method with inexact argument` to work, we can't only pick up `Slf4jLogsafeArgs` diagnostics, so we add `.matchAllDiagnostics()`.

With `.matchAllDiagnostics()` we need to add `@SuppressWarnings("deprecation")` in order to avoid diagnostics about `hasChildren` being deprecated.

aioobe committed:
Added support for logsafe Arg arrays in Slf4jLogsafeArgs.

aioobe committed:
Added support for logsafe Arg arrays in Slf4jLogsafeArgs.

aioobe committed:
Reduced cyclomatic complexity.

aioobe committed:
Reduced cyclomatic complexity.

aioobe committed:
Simplified logic. Vararg array argument now goes through regardless of type.

aioobe committed:
Simplified logic. Vararg array argument now goes through regardless of type.

carterkozak commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/Slf4jLogsafeArgs.java:
```diff
@@ -68,24 +73,27 @@ public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState
+        List<Integer> badArgIndices = IntStream.rangeClosed(startArgIndex, lastNonThrowableArgIndex)
+                .filter(i -> !ARG.matches(allArgs.get(i), state))
+                .boxed()
+                .collect(Collectors.toList());
```
Any chance we can avoid the stream here? This code is _super_ hot because we log a lot of things, one large internal project in particular benefits from any compile speedup we can provide!

aioobe replied:
Oh, I thought I spotted a nice simplification. Not my intention to slow it down. I'll revert to the previous for-each loop.

aioobe replied:
Reverted but had to factor it out to a method to pass the cyclomatic check.

carterkozak commented on baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/Slf4jLogsafeArgsTest.java:
```diff
@@ -47,6 +47,7 @@ private void test(String logArgs, String failingArgs) throws Exception {
                         "    public String getName() { return null; }",
                         "    public void add(Marker reference) {}",
                         "    public boolean remove(Marker reference) { return true; }",
+                        "    @SuppressWarnings(\"deprecation\")",
```
nice

carterkozak commented:
Description is a little out of date now, and it needs a changelog. I'm happy to merge this once those are in :-)

aioobe committed:
Reverted from Stream to for loop

aioobe committed:
Reverted from Stream to for loop

aioobe committed:
Added changelog entry.

aioobe committed:
Added changelog entry.

aldexis committed:
Improve gradle-baseline-java integration with IntelliJ import (#1411)

Adds the proper configuration files upon IntelliJ import of a gradle project for checkstyle and copyright.

This generates the following additional files:
- .idea/copyright/profiles_settings.xml
- an xml file under .idea/copyright/ per copyright file under .baseline/copyright
- .idea/checkstyle-idea.xml (and adds Checkstyle-IDEA to the external dependencies) if baseline-checkstyle is applied
- Either .idea/codeStyleSettings.xml or a .idea/codeStyles/ folder with the contents being copied from .baseline/idea
  - If .baseline/idea/codeStyles is present, it will copy its contents, otherwise, it will fall back to .baseline/idea/intellij-java-palantir-style.xml as currently
  - The fallback is using a legacy IntelliJ format and requires closing and reopening the project to be taken into account

svc-autorelease committed:
Autorelease 3.28.0

carterkozak commented:
I think this needs to have develop merged into it to pick up the unit-test-14 check

aioobe committed:
Merge branch 'fix-safe-args-warning' of github.com:aioobe/gradle-baseline into fix-safe-args-warning

ferozco commented:
👍

svc-autorelease commented:
Failed to load project - please reach out to #dev-foundry-infra or check Aries to debug
